shell/src/context_listener.rs

```diff
@@ -195,12 +205,54 @@ fn listen_protocol_events(
     Ok(())
 }
 
-fn perform_context_action(
+// returns hash of the merkle tree that action needs to be
+// applied on
+pub fn get_tree_hash(action: &ContextAction) -> Option<[u8;32]> {
+    match &action
+    {
+        ContextAction::Get {tree_hash, .. }
+        | ContextAction::Mem {tree_hash, .. }
+        | ContextAction::DirMem {tree_hash, ..}
+        | ContextAction::Set {tree_hash, ..}
+        | ContextAction::Copy {tree_hash, ..}
+        | ContextAction::Delete {tree_hash, ..}
+        | ContextAction::RemoveRecursively {tree_hash, ..}
+        | ContextAction::Commit {tree_hash, ..}
+        | ContextAction::Fold {tree_hash, ..} => {
+            let mut hash:[u8;32] = [0;32];
+            tree_hash.clone().reader().read_exact(& mut hash).unwrap();
+            Some(hash)
+        }
+        _ => {None}
```

Comment:
Whenever possible I would list everything explicitly here instead of using a wildcard, makes it easier to avoid mistakes related to updating code when a new tag is added to an enum. When using wildcards to match the rest of the enum possibilities we lost the exhaustiveness check from the compiler.